### PR TITLE
OG-851: Abandon Relay Servers free market 

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: packages/paymasters/test/ProxyDeployingPaymaster.test.ts [hunk 1]
```diff
@@ -319,8 +324,9 @@ contract('ProxyDeployingPaymaster', ([senderAddress, relayWorker, burnAddress])
       await paymaster.setRelayHub(testHub.address)
       const tx = await testHub.callPostRC(paymaster.address, context, gasUseWithoutPost, relayRequest.relayData)
       const gasUsedWithPost = gasUseWithoutPost + gasUsedByPost
+      const hubConfig = await testHub.getConfiguration()
       // Repeat on-chain calculation here for sanity
-      const actualEtherCharge = (100 + parseInt(relayRequest.relayData.pctRelayFee)) / 100 * gasUsedWithPost
+      const actualEtherCharge = (100 + parseInt(hubConfig.pctRelayFee.toString())) / 100 * gasUsedWithPost
```
Comment (drortirosh) [inline packages/paymasters/test/ProxyDeployingPaymaster.test.ts:329]: this calculation ignore baseRelayFee

File: packages/provider/src/KnownRelaysManager.ts [hunk 2]
```diff
@@ -109,11 +77,18 @@ export class KnownRelaysManager {
...
     sortedRelays[0] = Array.from(this.preferredRelayers)
-    sortedRelays[1] = await this._sortRelaysInternal(gsnTransactionDetails, this.allRelayers)
+    const relaysWithoutFailures = this.allRelayers.filter(it => {
+      return this.relayFailures.get(it.relayUrl) != null
+    })
+    const relaysWithFailures = this.allRelayers.filter(it => {
+      return this.relayFailures.get(it.relayUrl) == null
+    })
+    sortedRelays[1] = shuffle(relaysWithoutFailures)
+    sortedRelays[2] = shuffle(relaysWithFailures)
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/provider/src/KnownRelaysManager.ts:91]: on a call agreed that it is not supposed to be a major feature to 'rank' relays by failures as in 99.9% times there will be at least one successful relay

File: packages/dev/test/provider/RelaySelectionManager.test.ts [hunk 3]
```diff
@@ -101,17 +101,18 @@ contract('RelaySelectionManager', function (accounts) {
...
     it('should return the first relay to ping', async function () {
       stubGetNextSlice.returns([eventInfo])
-      stubRaceToSuccess
+      stubWaitForSuccess
         .onFirstCall()
-        .returns(Promise.resolve({ errors }))
+        .returns(Promise.resolve({ errors, results: [] }))
         .onSecondCall()
         .returns(Promise.resolve({
           winner,
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/dev/test/provider/RelaySelectionManager.test.ts:114]: yes.

File: packages/paymasters/test/ProxyDeployingPaymaster.test.ts (hunk 1 above)
Comment (forshtat) [reply packages/paymasters/test/ProxyDeployingPaymaster.test.ts:329]: this test is a bit abandoned, and baseFee is set to 0

File: packages/common/src/Utils.ts [hunk 4]
```diff
@@ -396,3 +396,83 @@ export function getERC165InterfaceID (abi: AbiItem[]): string {
...
+        ret.winner = ret.results[Math.floor(random() * ret.results.length)]
+      }
+      resolve(ret)
+    }
+
+    for (let i = 0; i < promises.length; i++) {
+      promises[i]
+        .then(result => {
+          ret.results.push(result)
+          if (ret.results.length + ret.errors.size === promises.length) {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/common/src/Utils.ts]: Fixed.

File: packages/provider/src/KnownRelaysManager.ts [hunk 5]
```diff
@@ -109,11 +77,18 @@ export class KnownRelaysManager {
...
-  async getRelaysSortedForTransaction (gsnTransactionDetails: GsnTransactionDetails): Promise<RelayInfoUrl[][]> {
+  async getRelaysShuffledForTransaction (): Promise<RelayInfoUrl[][]> {
     const sortedRelays: RelayInfoUrl[][] = []
     // preferred relays are copied as-is, unsorted (we don't have any info about them anyway to sort)
     sortedRelays[0] = Array.from(this.preferredRelayers)
-    sortedRelays[1] = await this._sortRelaysInternal(gsnTransactionDetails, this.allRelayers)
+    const relaysWithoutFailures = this.allRelayers.filter(it => {
+      return this.relayFailures.get(it.relayUrl) != null
+    })
+    const relaysWithFailures = this.allRelayers.filter(it => {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/provider/src/KnownRelaysManager.ts]: Fixed and added test.

**Review by drortirosh — APPROVED** [review]

**Pull request merged**